### PR TITLE
grid_map_core ROS2 port

### PR DESCRIPTION
**Pull request by Marwan99**
Related to #219

Port of grid_map_core package to ROS2.

- source and tests linted
- CMakeLists updated
- package.xml updated

### CODE REVIEW

**Comment by SteveMacenski** [comment]
Because this one is so large, can you highlight what you did? I imagine you updated the package xml and cmakelists then ran the ament_uncrustify automatic reformatter. What else? That will help me gloss over the auto-generated code changes and just focus in on the human changes

**Comment by Marwan99** [comment]
Major hand made changes:
- Swapping `int` with `std::size_t`  in for loops " looping variable" in loops where the compiler complained about comparing signed to unsigned.
- Resolving namespaces by hand.
- replacing `#pragma once` header guards, used the names suggested by the linter.

Will highlight the others in a review below for convince

File: grid_map_core/src/GridMap.cpp [hunk 1]
```diff
@@ -153,116 +173,137 @@ bool GridMap::erase(const std::string& layer) {
...
-      else
-        interpolationMethod = InterpolationMethods::INTER_NEAREST;
-    }
+        float value;
+        if (atPositionLinearInterpolated(layer, position, value)) {
+          return value;
+        } else {
+          interpolationMethod = InterpolationMethods::INTER_NEAREST;
+        }
+        [[fallthrough]];
```
Comment (Marwan99) [inline grid_map_core/src/GridMap.cpp:224]: using [fallthrough](https://en.cppreference.com/w/cpp/language/attributes/fallthrough) (C++17, is that okay?) I think the logic relies on not having break statements at the end of the cases, but the compiler was not happy with with that.

File: grid_map_core/src/GridMapMath.cpp [hunk 2]
```diff
@@ -463,105 +552,122 @@ bool incrementIndexForSubmap(Index& submapIndex, Index& index, const Index& subm
...
 
-void colorVectorToValue(const Eigen::Vector3i& colorVector, float& colorValue)
+void colorVectorToValue(const Eigen::Vector3i & colorVector, float & colorValue)
 {
-  unsigned long color = (colorVector(0) << 16) + (colorVector(1) << 8) + colorVector(2);
-  // cppcheck-suppress invalidPointerCast
-  colorValue = *reinterpret_cast<float*>(&color);
+  uint64_t color = (colorVector(0) << 16) + (colorVector(1) << 8) + colorVector(2);
+  colorValue = static_cast<float>(color);
 }
```
Comment (Marwan99) [inline grid_map_core/src/GridMapMath.cpp:665]: compiler was complaining about `*reinterpret_cast<float*>` replaced it with `static_cast<float>`

File: grid_map_core/src/Polygon.cpp [hunk 3]
```diff
@@ -6,49 +6,52 @@
...
 {
   int cross = 0;
-  for (int i = 0, j = vertices_.size() - 1; i < vertices_.size(); j = i++) {
-    if ( ((vertices_[i].y() > point.y()) != (vertices_[j].y() > point.y()))
-           && (point.x() < (vertices_[j].x() - vertices_[i].x()) * (point.y() - vertices_[i].y()) /
-            (vertices_[j].y() - vertices_[i].y()) + vertices_[i].x()) )
+  for (std::size_t i = 0, j = vertices_.size() - 1; i < vertices_.size(); j = i++) {
+    if ( ((vertices_[i].y() > point.y()) != (vertices_[j].y() > point.y())) &&
+      (point.x() < (vertices_[j].x() - vertices_[i].x()) * (point.y() - vertices_[i].y()) /
+      (vertices_[j].y() - vertices_[i].y()) + vertices_[i].x()) )
```
Comment (Marwan99) [inline grid_map_core/src/Polygon.cpp:41]: Example of replacing `int` with `std::size_t` in a for loop

File: grid_map_core/src/GridMap.cpp [hunk 4]
```diff
@@ -695,8 +813,9 @@ void GridMap::convertToDefaultStartIndex() {
   startIndex_.setZero();
 }
 
-Position GridMap::getClosestPositionInMap(const Position& position) const {
-  if (getSize().x() < 1u || getSize().y() < 1u) {
+Position GridMap::getClosestPositionInMap(const Position & position) const
+{
+  if (getSize().x() < 1 || getSize().y() < 1) {
```
Comment (Marwan99) [inline grid_map_core/src/GridMap.cpp:818]: Example of resolving a complain about comparing signed to unsigned.

File: grid_map_core/test/GridMapIteratorTest.cpp [hunk 5]
```diff
@@ -21,43 +18,48 @@
...
+  auto & data = map["layer"];
   unsigned int i = 0;
   for (; !iterator.isPastEnd(); ++iterator, ++i) {
     data(iterator.getLinearIndex()) = 1.0;
     EXPECT_EQ(i, iterator.getLinearIndex());
     EXPECT_FALSE(iterator.isPastEnd());
   }
 
-  EXPECT_EQ(40, i);
+  EXPECT_EQ(40u, i);
```
Comment (Marwan99) [inline grid_map_core/test/GridMapIteratorTest.cpp:62]: Another example of resolving compiler complains about comparing signed to unsigned. This error was very common in the test files.

**Comment by Marwan99** [comment]
@SteveMacenski If there is something else I can do to make it slightly easier please let me know. Apologies for the big PR.

**Comment by maximilianwulf** [comment]
@SteveMacenski and @Marwan99 I think there are many improvements on the source that would also help in the ROS1 version. With newer packages, we enforce clang, which is not added. Would it make sense to update the ROS1 version first and then integrate it into ROS2?

**Comment by YoshuaNava** [comment]
@maximilianwulf Does it make sense to split this PR into small ones?

File: grid_map_core/src/GridMap.cpp (hunk 1 above)
Comment (SteveMacenski) [reply grid_map_core/src/GridMap.cpp:224]: You should use C++14 unless absolutely required to use C++17 - ROS2 Foxy is under C++14.

File: grid_map_core/src/GridMapMath.cpp (hunk 2 above)
Comment (SteveMacenski) [reply grid_map_core/src/GridMapMath.cpp:665]: Fine by me :+1:

**Comment by SteveMacenski** [comment]
@maximilianwulf most of these updates were because of the picky compiler flags we set in the common packages and because of the ROS2 linters. If things are working in ROS1, I wouldn't go back and try to make all these changes since it would be time consuming and for little tangible benefit. Something like a C style cast is totally valid, but the ROS2 style guide complains about it because it would rather use a std cast. 

@YoshuaNava for this one, potentially, but they'll all be massive. >50% of this work is done from automated linter reformatting.  If we were to split it up, it would make sense to split it up into 2 PRs
- Automated linting - just autoapprove assuming compiles and passes tests
- Hand linting - the things we would need to spend a little time just sanity checking.

File: grid_map_core/test/EllipseIteratorTest.cpp [hunk 6]
```diff
@@ -21,16 +18,19 @@
 // Vector
 #include <vector>
 
-using namespace std;
-using namespace Eigen;
-using namespace grid_map;
+#include "grid_map_core/iterators/EllipseIterator.hpp"
+#include "grid_map_core/GridMap.hpp"
+
+// using namespace std;
```
Comment (maximilianwulf) [inline grid_map_core/test/EllipseIteratorTest.cpp:24]: Delete.

File: grid_map_core/test/GridMapIteratorTest.cpp [hunk 7]
```diff
@@ -21,43 +18,48 @@
 // Vector
 #include <vector>
 
-using namespace std;
-using namespace grid_map;
+#include "grid_map_core/iterators/GridMapIterator.hpp"
+#include "grid_map_core/GridMap.hpp"
+
+// using namespace std;
```
Comment (maximilianwulf) [inline grid_map_core/test/GridMapIteratorTest.cpp:24]: Delete

File: grid_map_core/test/GridMapMathTest.cpp [hunk 8]
```diff
@@ -20,231 +18,380 @@
 // Vector
 #include <vector>
 
-using namespace std;
-using namespace grid_map;
+#include <limits>
+
+#include "grid_map_core/GridMapMath.hpp"
+
+// using namespace std;
```
Comment (maximilianwulf) [inline grid_map_core/test/GridMapMathTest.cpp:25]: Delte

File: grid_map_core/test/LineIteratorTest.cpp [hunk 9]
```diff
@@ -6,24 +6,27 @@
...
 #include <gtest/gtest.h>
 
 // Limits
 #include <cfloat>
 
-using namespace grid_map;
+#include "grid_map_core/iterators/LineIterator.hpp"
+#include "grid_map_core/GridMap.hpp"
+
+// using namespace grid_m ap;
```
Comment (maximilianwulf) [inline grid_map_core/test/LineIteratorTest.cpp:18]: Delete

File: grid_map_core/test/PolygonIteratorTest.cpp [hunk 10]
```diff
@@ -6,40 +6,43 @@
...
-using namespace std;
-using namespace Eigen;
-using namespace grid_map;
+#include <string>
+
+#include "grid_map_core/iterators/PolygonIterator.hpp"
+#include "grid_map_core/GridMap.hpp"
+#include "grid_map_core/Polygon.hpp"
+
+// using namespace std;
```
Comment (maximilianwulf) [inline grid_map_core/test/PolygonIteratorTest.cpp:27]: Delete

File: grid_map_core/test/PolygonTest.cpp [hunk 11]
```diff
@@ -6,59 +6,61 @@
...
 #include <Eigen/Dense>
 
-using namespace std;
-using namespace Eigen;
-using namespace grid_map;
+#include <vector>
+
+#include "grid_map_core/Polygon.hpp"
+
+// using namespace std;
```
Comment (maximilianwulf) [inline grid_map_core/test/PolygonTest.cpp:20]: As above.

**Comment by maximilianwulf** [comment]
> @maximilianwulf Does it make sense to split this PR into small ones?

In general I would say yes, but most of the changes are only to the layout. As long as the unit tests pass I think we are on the safe side, @Marwan99?

File: grid_map_core/test/GridMapTest.cpp [hunk 12]
```diff
@@ -6,24 +6,27 @@
...
 #include <math.h>
 
-using namespace std;
-using namespace grid_map;
+#include <vector>
+#include <string>
+
+#include "grid_map_core/GridMap.hpp"
+
+// using namespace std;
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply grid_map_core/test/GridMapTest.cpp]: Just kept them as a reference while resolving namespaces, they are not needed now.

File: grid_map_core/src/Polygon.cpp [hunk 13]
```diff
@@ -220,13 +225,15 @@ bool Polygon::offsetInward(const double margin)
   return true;
 }
 
-std::vector<Polygon> Polygon::triangulate(const TriangulationMethods& method) const
+std::vector<Polygon> Polygon::triangulate(const TriangulationMethods & method) const
 {
-  // TODO Add more triangulation methods.
+  // TODO(needs_assignment): Add more triangulation methods.
   // https://en.wikipedia.org/wiki/Polygon_triangulation
+  (void)(method);  // method parameter unused, should be removed once used.
```
[the remark this answers is not in the document]
Comment (Marwan99) [reply grid_map_core/src/Polygon.cpp:232]: Sure, will add it. A TODO needs to be assigned to someone (linter requirement), who should I put down? 

Also there are another 5 instances where parameters are not used and the same work around `(void)(param);` was used, I can list them if needed.  

On a similar note, in the existing TODOs I have added "someone" as a place holder, should that be updated as well?

[the remark this answers is not in the document]
Comment (maximilianwulf) [reply grid_map_core/src/Polygon.cpp:232]: Yes, add a placeholder.

**Comment by SteveMacenski** [comment]
I'm good to merge when @Marwan99 CI passes (thanks for that!) and @maximilianwulf approves

**Comment by Marwan99** [comment]
@SteveMacenski I need some help!
- Not sure why I am getting `readability/mixedcase` linting error since the whole message is upercase (last 2 failures). Any suggestions?
- Despite having `set(ament_cmake_copyright_FOUND TRUE)` the linter is still complaining about the lack of copyright messages. Do I have it in the wrong place? 

On another note the previous compiling error (before the last commit), was not happening locally on my machine and it was compiling perfectly fine, any suggestions why this might have happened?

**Comment by SteveMacenski** [comment]
The readibility / mixed cases is in the cmake file, I think it just wants you to rename things on L4, 13. I think the legal one is from cpplint not the ament_copyright checker. Maybe there's a way to disable the specific `legal/copyright [5]` check in `cpplint`?

File: grid_map_core/CMakeLists.txt [hunk 14]
```diff
@@ -81,50 +55,61 @@ target_link_libraries(${PROJECT_NAME}
...
-  test/EllipseIteratorTest.cpp
-  test/SubmapIteratorTest.cpp
-  test/PolygonIteratorTest.cpp
-  test/PolygonTest.cpp
-  test/EigenPluginsTest.cpp
-  test/SpiralIteratorTest.cpp
-  test/SlidingWindowIteratorTest.cpp
+if(BUILD_TESTING)
+  find_package(ament_lint_auto REQUIRED)
+  set(ament_cmake_copyright_FOUND TRUE)
```
Comment (naiveHobo) [inline grid_map_core/CMakeLists.txt:82]:
```suggestion
  find_package(ament_lint_auto QUIET)
  if(ament_lint_auto_FOUND)
    # exclude copyright checks
    list(APPEND AMENT_LINT_AUTO_EXCLUDE
      ament_cmake_cpplint
      ament_cmake_copyright
    )
    ament_lint_auto_find_test_dependencies()

    # run cpplint without copyright filter
    find_package(ament_cmake_cpplint)
    ament_cpplint(
      FILTERS -legal/copyright
    )
  endif()
```

Comment (naiveHobo) [reply grid_map_core/CMakeLists.txt:82]: This should disable all the copyright checks.

Comment (Marwan99) [reply grid_map_core/CMakeLists.txt:82]: Thank you so much @naiveHobo! I found the FILTERS but I was struggling to get it working with ament_lint_auto.

File: grid_map_core/CMakeLists.txt [hunk 15]
```diff
@@ -79,7 +79,21 @@ install(
...
   find_package(ament_lint_auto REQUIRED)
-  set(ament_cmake_copyright_FOUND TRUE)
+  find_package(ament_lint_auto QUIET)
+  if(ament_lint_auto_FOUND)
+    # exclude copyright checks
+    list(APPEND AMENT_LINT_AUTO_EXCLUDE
+      ament_cmake_cpplint
+      ament_cmake_copyright
+    )
+    ament_lint_auto_find_test_dependencies()
```
Comment (SteveMacenski) [inline grid_map_core/CMakeLists.txt:89]: You call this 2x now

**Comment by SteveMacenski** [comment]
Looks like 1 more linter error - also did your lines you added remove cpplint from core overall? I see one line where you exclude the copyright checks but it looks like you might also just turn it off

**Comment by naiveHobo** [comment]
@SteveMacenski it removes copyright check (which is independent of cpplint) as well as cpplint (due to the separate `legal/copyright` checks) from `ament_lint_auto`. And cpplint is then run separately with the `legal/copyright' filter disabled.

**Comment by SteveMacenski** [comment]
Ah ok, that makes sense. Can we make a note in the cmakelists to that effect so that's clear. I could see someone a year from now wanting to refactor that thinking we did something really silly

**Comment by Marwan99** [comment]
If I'm understanding the error correctly, I can't think of a way to resolve the current cppcheck problem without breaking the API for the `float & GridMap::at` method. Is there a less destructive way of sorting this out?

**Comment by SteveMacenski** [comment]
Please link to this function definition here

**Comment by Marwan99** [comment]
https://github.com/Marwan99/grid_map/blob/db19acf8e2afb4ad01173415210f2fa07367974f/grid_map_core/src/GridMap.cpp#L232-L239

**Comment by SteveMacenski** [comment]
mhm good question, that's a real issue. `at` returns a reference to the value in the map / unordered map, but then you access the reference and return some float value stored in the matrix. I googled around and it looks like the `operator()` in the `Eigen::MatrixXf` returns a `float &` so its also a reference to the class variable.

Therefore, I believe this is safe and can be given a `// NOLINT` (if that works for cppcheck as well) to be ignored. @maximilianwulf please verify that that makes sense.

**Comment by naiveHobo** [comment]
Can this me merged quickly? I have ported `grid_map_costmap_2d` and `grid_map_octomap` to ros2 branching off this version. Would be great if we can close this asap.

**Comment by SteveMacenski** [comment]
Need @maximilianwulf to give it another once over and approve

File: grid_map_core/src/GridMapMath.cpp [hunk 16]
```diff
@@ -463,105 +552,122 @@ bool incrementIndexForSubmap(Index& submapIndex, Index& index, const Index& subm
...
+    return Index(
+      static_cast<int>(linearIndex) % bufferSize(
+        0), static_cast<int>(linearIndex) / bufferSize(0));
+  }
+  return Index(
+    static_cast<int>(linearIndex) / bufferSize(
+      1), static_cast<int>(linearIndex) % bufferSize(1));
+}
+
+// void getIndicesForRegion(
```
[the remark this answers is not in the document]
Comment (maximilianwulf) [reply grid_map_core/src/GridMapMath.cpp:610]: In general, we should just delete it. Can you delete it?

File: grid_map_core/CMakeLists.txt (hunk 15 above)
Comment (maximilianwulf) [reply grid_map_core/CMakeLists.txt:95]: @Marwan99 is there a reason for it?

**Comment by maximilianwulf** [comment]
> Therefore, I believe this is safe and can be given a // NOLINT (if that works for cppcheck as well) to be ignored. @maximilianwulf please verify that that makes sense.

Did you manage to fix it without linting? I didn't find the corresponding change. 

Besides that, LGTM. Good job!

**Comment by Marwan99** [comment]
@maximilianwulf This error was from cpp_check (which performs static code analysis, which is done with the linting checks), this was fixed by adding a comment telling the checker to ignore it, something similar to `// NOLINT`. So the API **was not** modified. You can find the fix in 507899f.

**Comment by SteveMacenski** [comment]
just waiting on your OK @maximilianwulf

**Comment by maximilianwulf** [comment]
Okay, got it. Thanks a lot.

**Comment by SteveMacenski** [comment]
In@

**Pull request merged**